### PR TITLE
Shard datasets for distributed training.

### PR DESCRIPTION
**Pull request by jysohn23**
Added torch.utils.data.distributed.DistributedSampler usage to our
MNIST, CIFAR, and IMAGENET examples and tested for training accuracy
improvements with larger LR and that each shard had a different loss
(data parallelism).

Verified no performance regressions on v3-8/n1-standard-64:
* Baseline (no sharding):
https://gist.github.com/jysohn23/8077aab85474ef3e7bfcb54851280d8c
* Sharding (single-shard):
https://gist.github.com/jysohn23/e0900d888534598f06b2eec3da598d14

Tested on v3-32 & 4 x n1-standard-64
(May suffer from lower accuracy since larger global batch sizes; all
use real data)
MNIST: https://gist.github.com/jysohn23/0ba2d58c2da553828c29461e89f8d1da
CIFAR: https://gist.github.com/jysohn23/0b6b57e50007e57189b35032ed654a4f
IMAGENET (ResNet50):
https://gist.github.com/fb5a50abea01b273fc55599ce6994559

### CODE REVIEW

File: test/test_train_cifar.py [hunk 1]
```diff
@@ -122,26 +122,31 @@ def train_cifar():
...
-        shuffle=True,
-        num_workers=FLAGS.num_workers)
-
-    testset = torchvision.datasets.CIFAR10(
+    test_dataset = torchvision.datasets.CIFAR10(
         root=FLAGS.datadir,
         train=False,
         download=True,
         transform=transform_test)
+    train_sampler = torch.utils.data.distributed.DistributedSampler(
```
Comment (dlibenzi) [inline test/test_train_cifar.py]: I'd feel better if this is done if world_size > 1.
IOW, pass `sampler=None` (or whatever default) if world_size==1.

File: torch_xla_py/xla_model.py [hunk 2]
```diff
@@ -83,6 +83,14 @@ def get_xla_supported_devices(devkind=None, max_devices=None):
       return kind_devices[:max_devices] if max_devices else kind_devices
 
 
+def xrt_world_size():
+  return int(os.getenv('XRT_NUM_WORKERS', 1))
```
Comment (dlibenzi) [inline torch_xla_py/xla_model.py]: Actually, I have a second thought about this.
We should be (in the `xla_dist.py` case) exporting explicit variables about "ordinal" and "world size".
Maybe eve using the same of pytorch?

Reason is, with the new multiprocessing, there will be one shard per core, for the concept of worker becomes a little loose.
In that case, a "worker" is meant as a process-group handling one TPU device, and each of the "forks" will have a device assigned.

Comment (jysohn23) [reply torch_xla_py/xla_model.py]: Ok, let me see if we're on the same page.

Let's say we have 1:1 setup between Compute VM and TPU VM, and the TPU is a `v3-32`. Then in the multiprocessing world, we'd have 4 workers (aka. process groups) but 32 separate processes/shards. So in that case we would have:
* `XRT_LOCAL_WORKER=c_tpu_worker:0...3`
* `XRT_SHARD_ORDINAL=0...31`
* `XRT_SHARD_WORLD_SIZE=32`

Is that correct?

File: test/test_train_cifar.py (hunk 1 above)
Comment (jysohn23) [reply test/test_train_cifar.py]: Done. Tested on all 3 datasets and v3-8/v3-32.

File: torch_xla_py/xla_model.py (hunk 2 above)
Comment (jysohn23) [reply torch_xla_py/xla_model.py]: Also, during the meantime while we don't have multiprocessing, I'll set `XRT_SHARD_ORDINAL=0...3` and `XRT_SHARD_WORLD_SIZE=4` since 1 shard = 1 worker = 1 process until we have multiprocessing.

Comment (dlibenzi) [reply torch_xla_py/xla_model.py]: For now, you only set XRT_SHARD_ORDINAL=0..3 and XRT_SHARD_WORLD_SIZE=4.
The multiprocessing will have to rewrite those.
We will have to support (at least for a while) bot the multi-thread and multi-processing interfaces anyway.

Comment (jysohn23) [reply torch_xla_py/xla_model.py]: Done.
